client/www/app/core/dataservice.js

```diff
@@ -102,15 +110,87 @@
       });
     }
     
-    // NOTE: moment is saved to moments table first, then added to memento
-    function saveMoment(obj) {
-      obj.ID = momentsSize() + 1;
+    // uploads moment content to S3, then saves moment to database
+    function saveMoment(moment) {
+      return uploadItems(moment.content)
+        .then(function(momentContent) {
+          // S3 urls added to moment content
+          moment.content = momentContent
+
+          return $http.post('http://mementos.ngrok.com/api/1/moments', moment)
+            .success(function(momentID, status, headers, config) {
+              console.log('Successful saving moment');
+              return momentID;
+            })
+            .error(function(data, status, headers, config) {
+              console.log('There was an error saving moment');
+            });  
+        })
+        .catch(function(err) {
+          console.log('There was an error retrieving all moment content');
+        });
+      
+      /*NOTE: dummy data code to be removed*/
+      /*obj.ID = momentsSize() + 1;
       
       moments.push(obj);
 
       return $q(function(resolve, reject) {
         resolve(obj.ID);
-      });
+      });*/
+    }
+    
+    // sends each moment item to uploadItem and notifies saveMoment when uploads are complete
+    function uploadItems (items) {
+      var unfinished = items.length;
+      var running = 0;
+      var deferred = $q.defer();
+      var momentItems = [];
+      var i;
+      
+      // NOTE: async implementation may not work 100% of time
+      for (i = 0; i < items.length; i++) {
+        (function(index) {
+          running++;
+          uploadItem(items[index])
+            .then(function(S3Data) {
+              momentItems.push({type: S3Data.config.url.split('type=')[1], url: S3Data.data.url});
+              running--;
+              unfinished--;
+
+              if(running === 0 && unfinished === 0) {
+                deferred.resolve(momentItems);
+              }
+            })
+            .catch(function(err) {
+              console.log('There was an error uploading to S3', err);
+            });
+        }(i));
```

Comment:
In this case, it's best if we send the contents instead of the index. You can call the IIFEE with items[i] instead of the index. Looks much cleaner